File: ios/Cookpit/Cookpit/SearchViewModel.swift
//
//  SearchViewModel.swift
//  Cookpit
//
//  Created by Kittinun Vantasin on 6/1/16.
//  Copyright © 2016 Cookpit. All rights reserved.
//

import Foundation
import RxSwift

enum SearchViewModelCommand {

  case SetSearchItems(items: [CPSearchDetailViewData])
  case SetRecentItems(items: [String])
  
}

struct SearchViewModel {

  let searchItems: [CPSearchDetailViewData]
  let recentItems: [String]
  
  func executeCommand(command: SearchViewModelCommand) -> SearchViewModel {
    switch command {
    case let .SetSearchItems(items):
        return SearchViewModel(searchItems: items, recentItems: recentItems)
    case let .SetRecentItems(items):
        return SearchViewModel(searchItems: searchItems, recentItems: items)
    }
  }
  
}

File: android/Cookpit/app/src/main/kotlin/com/github/kittinunf/cookpit/search/SearchViewModel.kt
package com.github.kittinunf.cookpit.search

import com.github.kittinunf.cookpit.SearchDetailViewData

sealed class SearchViewModelCommand {

    class SetRecentItems(val items: List<String>) : SearchViewModelCommand()
    class SetSearchItems(val items: List<SearchDetailViewData>) : SearchViewModelCommand()

}

data class SearchViewModel(val recentItems: List<String> = listOf(), val searchResults: List<SearchDetailViewData> = listOf()) {

    fun executeCommand(command: SearchViewModelCommand): SearchViewModel {
        when(command) {
            is SearchViewModelCommand.SetRecentItems -> {
                return SearchViewModel(command.items, searchResults)
            }
            is SearchViewModelCommand.SetSearchItems -> {
                return SearchViewModel(recentItems, command.items)
            }
        }
    }

}
